### ISSUE TITLE
Only allow users to edit and delete their own queries, unless admin

### ISSUE THREAD

**Issue by vweevers** [rickbergfalk/sqlpad#145]
Would you be open to a PR? It can be an opt-in feature, if need be.

**Comment by rickbergfalk** [comment]
Definitely! I'm not using this in a team setting at all, so I'm all for it being shaped by real-world use :)

**Comment by vweevers** [comment]
@rickbergfalk your code hints at a future "viewer" role. What are/were your plans with that?

**Comment by rickbergfalk** [comment]
It was going to be a role that could view and execute the queries as written by others, without the ability to modify the query prior to execution. I hadn't decided on whether the query editor UI would be presented to the user as is but locked down, or if It would be something different. 

Since the viewer role would only consume query results, I was thinking the query could run on open, kind of like how the grid-only/chart-only pages work. Maybe it would default to the chart view if a chart was configured. The actual SQL run could be viewable for reference, but it wouldn't be a primary focus when opening the query. 

I haven't really thought about this beyond that. I haven't had much time lately to flesh out my remaining sqlpad ideas. 

Also worth nothing - All my sqlpad has been as a solo user, so I'm not always entirely sure what's useful for a team and what isn't.

If this is something you are interested in feel free to make it however feel would be beneficial. I'm not invested in any particular vision of what this might be like in practice :)

**Comment by lexadavid** [comment]
This feature would be amazing. My team started using SQLPad heavily and I worry that anybody can delete other people's queries. 
If only admin can do it, the problem is solved.

Thanks for great tool @rickbergfalk!

**Comment by vweevers** [comment]
FIY I've implemented an ACL for SQLPad, with which one can grant or revoke a `principal` one or more `actions` to a `resource`. It looks like this:

```js
Rule.grant('role:admin', '*', '*')
Rule.grant('role:editor', 'query', ['create', 'read'])

// Can also add rules for specific users and resources
Rule.revoke('user:<userid>', 'query:<query id>', ['update']);
```

While not feature-complete, we've been using it in our team for about 1.5 months now, without problems. Hopefully I can find some time soon to push this upstream.

**Comment by rickbergfalk** [comment]
Sorry for the delay in getting to this, but thanks for for the kind words @lexadavid! 

And @vweevers - the ACL would be amazing to add in. I'm all for pulling in your additions, especially given it is seeing real usage with your team.

**Comment by davidmwilliams** [comment]
Hi @vweevers, were you able to push your mods?

**Comment by rickbergfalk** [comment]
@vweevers I assume with the passing of time and all the code changes that it is safe to say the ACL won't be merged upstream?

Just curious... with the other request about assigning connections to users I'm giving more thought to implementing some sort of RBAC/ACL model...

**Comment by vweevers** [comment]
> I assume with the passing of time and all the code changes that it is safe to say the ACL won't be merged upstream?

Yea. I wish it were otherwise, but I don't have the capacity for it.

**Comment by rickbergfalk** [comment]
@vweevers no worries :)

**Comment by rickbergfalk** [comment]
Closing because realistically not something I'm going to be working on. I find myself using SQLPad less and less lately and definitely not with a team, and I'm hesitant to add features I personally won't be using.

**Comment by rickbergfalk** [comment]
Re-opening as this is something I'm going to be working on very soon. 

Existing queries in existing SQLPad environments will be migrated to have an entry saved sharing the query with everyone to maintain today's behavior. 

Going forward though, new queries will start out private (open to having this configurable, but would like to avoid it if possible)
